python/ray/serve/_private/proxy_state.py

```diff
@@ -194,100 +192,97 @@ def actor_handle(self) -> ActorHandle:
-                ray.get(finished[0])
-                return ProxyWrapperCallStatus.FINISHED_SUCCEED
-            else:
-                return ProxyWrapperCallStatus.PENDING
-        except RayActorError:
-            return ProxyWrapperCallStatus.FINISHED_FAILED
+                return False
+            finally:
+                self._ready_check_future = None
+        else:
+            return None
 
-    def is_drained(self) -> ProxyWrapperCallStatus:
-        """Return whether the proxy actor is drained or not.
+    def is_healthy(self, timeout_s: float) -> Optional[bool]:
+        if self._health_check_future is None:
+            self._health_check_future = wrap_as_future(
+                self._actor_handle.check_health.remote(), timeout_s=timeout_s
+            )
+            return None
+        elif self._health_check_future.done():
+            try:
+                # NOTE: Since `check_health` method is responding with nothing, sole
+                #       purpose of fetching the result is to extract any potential
+                #       exceptions
+                self._health_check_future.result()
+                return True
+            except Exception as e:
+                if isinstance(e, TimeoutError):
+                    logger.warning(
+                        f"Didn't receive health check response for proxy"
+                        f" on {self._node_id} after {timeout_s}s."
+                    )
+                else:
+                    logger.error(
+                        f"Unexpected error invoking health check for proxy "
+                        f"on {self._node_id}",
+                        exc_info=e,
+                    )
 
-        If the ongoing drained check is finished, and the value can be retrieved,
-        reset _is_drained_obj_ref to ensure drained check is finished and return
-        FINISHED_SUCCEED status. If the ongoing ready check is not finished,
-        return PENDING status.
-        """
-        finished, _ = ray.wait([self._is_drained_obj_ref], timeout=0)
-        if finished:
-            self._is_drained_obj_ref = None
-            ray.get(finished[0])
-            return ProxyWrapperCallStatus.FINISHED_SUCCEED
+                return False
+            finally:
+                self._health_check_future = None
         else:
-            return ProxyWrapperCallStatus.PENDING
+            return None
+
+    def is_drained(
+        self, timeout_s: float = PROXY_HEALTH_CHECK_TIMEOUT_S
+    ) -> Optional[bool]:
+        if self._drained_check_future is None:
+            self._drained_check_future = wrap_as_future(
+                self._actor_handle.is_drained.remote(),
+                # NOTE: We use the same timeout as for health-checking
+                timeout_s=timeout_s,
+            )
+            return None
+        elif self._drained_check_future.done():
+            try:
+                # NOTE: Since `check_health` method is responding with nothing, sole
```

Comment:
I share the sentiment, but it's a balance we need to strike. In this case the sequences are practically identical but they also not easy to abstract b/c they are stateful and b/c of that abstraction actually brings more complexity for less value than non-abstracted version

Reply:
oh, no. I think it has to be it's own function (also tried to refactor all checks pervious and it's getting very ugly if we want to generalize it). This comment is just noting the comment should be updated since this one is for drain check XD

Reply:
Ah, good catch!